### DISCUSSION TITLE
Deploying Custom Clients (GPO)

### DISCUSSION THREAD

**Discussion by Koruudo** [rustdesk/rustdesk#7547] [General]
I absolutely LOVE the new custom clients feature for Server Pro, however deploying these clients is difficult.

First, after it builds, it creates an exe. Once you launch that EXE, it extracts the program files to your user's AppData\Local folder. Then, once the app opens and you click the **Install** button to install it, it then creates MORE files under AppData\Local\Temp, such as batch scripts for installing the app.

The difficulty here is all these layers make it fuzzy on how I could deploy this through Group Policy... maybe a script that runs the first exe, then runs the next one with the "--silent-install wait" parameter? I tried that and it failed.

How do we go about deploying custom clients automatically, with tools such as Group Policy?

**Comment by rustdesk** [comment]
Really? Custom client works the same way as our stable release. Many users have deployed the stable release.

**Reply by Koruudo** [reply]
So do I use the standard deployment script but use the custom client exe instead?

The power shell script contains things like setting the custom configuration key and password, but the custom client already does that. So I wasn't sure if this was as simple as calling the custom client exe and passing the "--silent-install" argument

**Reply by rustdesk** [reply]
Frankly, I have never run these script, all of these have been written by contributors. What we have done is adding command line support for these script when contributors asked for.

**Reply by Koruudo** [reply]
Hmmm, I guess I'll have to tinker with this some more.

Can you confirm if there are arguments that can be passed to the custom client EXE? Or do I have to manually run it interactively on each computer?

**Reply by rustdesk** [reply]
custom EXE is almost the same with stable release (technically they are the same EXE, we only change the icon resource, and sign again). We do not add or remove code. custom EXE is just some preset config based on stable release.

**Reply by rustdesk** [reply]
Sorry, there is one diff, custom EXE is based on nightly release, not the stable release, because current stable release does not support customization.

**Comment by darnasii** [comment]
If the generator generated the MSI file, then the problem would be solved :)

**Reply by rustdesk** [reply]
Agree

**Reply by darnasii** [reply]
Is there any chance to get this feature in near future?

**Reply by rustdesk** [reply]
Yes, you will see soon

**Comment by Koruudo** [comment]
Update: I used a program called AdvancedRun to run powershell as SYSTEM and execute the EXE with `--silent-install wait` passed as an argument and it worked successfully, so theoretically this can be pushed out through Group Policy as a Startup Script and install fine.

My only problem I've found is, not sure if this is isolated to me, but after the installation I cannot open the app by clicking the systray icon, I get the following error:
![Screenshot 2024-03-29 083132](https://github.com/rustdesk/rustdesk/assets/25572885/6c6a2b84-d06b-4985-8a99-dc05b0bbbf7f)

**Reply by rustdesk** [reply]
We have a fix for this, https://github.com/rustdesk/rustdesk/pull/7540, some one reported this 2 days ago. Not release yet.

**Reply by rustdesk** [reply]
Thanks for your effort. To make life easier, we will generate MSI.

**Reply by Koruudo** [reply]
My second time pushing the install on another machine didn't seem to do it so that issue may be a hit or miss, but not critical issue.

**Reply by Koruudo** [reply]
This would be wonderful, thank you! :)

**Comment by Koruudo** [comment]
> We have a fix for this, #7540, some one reported this 2 days ago. Not release yet.



> Thanks for your effort. To make life easier, we will generate MSI.

Answered, on the same page. Closing.

**Comment by rustdesk** [comment]
> Thanks for sending that link over, I discovered that Windows was blocking the exe installing.
> 
> In a business environment, the running of the downloaded exe as admin doesn't pass that privilege to the extracted exe located in the AppData\Local directory.
> 
> Running the AppData exe allowed the installation to run - as mentioned an MSI install would probably get round this!

**Comment by rustdesk** [comment]
Ready now, https://twitter.com/rustdesk/status/1782250852574769663

**Reply by My1** [reply]
is it allowed to push the android custom client into the Playstore? would be kinda useful especially as enabling assistive stuff for apk installs has been knackered by Android for a while.

**Reply by Koruudo** [reply]
> is it allowed to push the android custom client into the Playstore? would be kinda useful especially as enabling assistive stuff for apk installs has been knackered by Android for a while.

I don't see pushing it into the public play store, but it might be possible with the Google Admin Console if you have Google workspace. But last time I tried privately deploying a custom APK I got an error and couldn't.

**Reply by My1** [reply]
Well rustdesk would obviously have to decide on that, but i wouldn't see why not especially as asking customers to enter a server and a key is way too annoying

(and having the server gated in a way that you have to login before controlling, would reduce the whole "this app is used for scam" issue)

**Reply by Koruudo** [reply]
Publishing it to the public playstore is basically impossible, as every custom client anybody makes would have to be submitted to RustDesk, RustDesk publishing it as its **own** app, and very quickly you'll have hundreds of different RustDesk apps on the playstore; it doesn't make any sense that way.

If you have a custom android client, the only feasible way of deploying it is through Google Admin Console as a "Private Android App" - like I said earlier, you can upload APKs that way and deploy them. I'm going to try making a custom Android client and try deploying that. Last time I just tried deploying the stock RustDesk android app through the admin console, I got the following error:
![image](https://github.com/rustdesk/rustdesk/assets/25572885/80b61dcf-06e1-4d1b-bf3c-86577b9185eb)

...which is weird because it doesn't exist on the playstore, so I'm not sure why it's saying it's "preinstalled". 😅

Maybe @rustdesk can look into this sometime. I'll try a custom client and update on this thread how that goes.

**Reply by My1** [reply]
deploying stock sure wont work, true, I would also have assumed custom rustdesks not being called "rustdesk" but rather "Company X Support Tool" or whatever and having its own Package name.

supplying Rustdesk through a private Playstore with Google for work would mean you can only use that with people whose phones you manage, you cannot really offer any public support.

and offering public support with your self hosted server seems kinda unfeasible.

**Reply by Koruudo** [reply]
Publishing your own custom client to the playstore under your company name would be a large load on RustDesk's shoulders, though, unless you manually create your own developer account and try publishing to the play store yourself. There's too many variables in the process of publishing an app to the playstore to make it practical to be done for every single custom client customers would make.

**Reply by My1** [reply]
> Publishing your own custom client to the playstore under your company name would be a large load on RustDesk's shoulders, though, unless you manually create your own developer account and try publishing to the play store yourself. There's too many variables in the process of publishing an app to the playstore to make it practical to be done for every single custom client customers would make.

of course publishing yourself. that was the whole premise. especially as custom clients seem to come from the server rather than a web portal which means they are (in case of windows) not signed which has its own problems but on Android you dont need to pay like 200€ a year or whatever for a code signing. you pay once for a Playstore account and can upload.

the only question is whether just pushing one in would be allowed per license and all, especially as I dunno the pro license (or heck even manual forks of the open version that pin to your server and all)

**Reply by Koruudo** [reply]
Yeah, that would be at RustDesk's discretion and in accordance with their open-source license.

**Comment by Koruudo** [comment]
> Ready now, https://twitter.com/rustdesk/status/1782250852574769663

You guys are the best!!!! 💖 Testing the custom client MSI now :D
